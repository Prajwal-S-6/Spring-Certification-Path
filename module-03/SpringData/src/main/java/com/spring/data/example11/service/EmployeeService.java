package com.spring.data.example11.service;


import com.spring.data.example11.dao.EmployeeDao;
import com.spring.data.example11.ds.Employee;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.sql.Date;

@Service
public class EmployeeService {

    @Autowired
    private EmployeeDao employeeDao;

    public void saveEmployeeWithoutTransaction() {
        System.out.println("Saving employee without transaction...");
        employeeDao.saveEmployee(new Employee(1, "John", "Doe", "John.Doe@corp.com", "555-55-55", Date.valueOf("2019-06-05"), 70000));
    }

    @Transactional
    public void saveEmployeeInTransaction() {
        System.out.println("Saving employee with transaction...");
        employeeDao.saveEmployee(new Employee(1, "John", "Doe", "John.Doe@corp.com", "555-55-55", Date.valueOf("2019-06-05"), 70000));
    }
}
