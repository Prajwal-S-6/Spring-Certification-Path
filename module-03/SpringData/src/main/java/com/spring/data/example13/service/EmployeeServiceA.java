package com.spring.data.example13.service;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Transactional
public class EmployeeServiceA {

    public void operationA() {
    }
}
