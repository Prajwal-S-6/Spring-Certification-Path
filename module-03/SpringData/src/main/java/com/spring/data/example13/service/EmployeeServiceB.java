package com.spring.data.example13.service;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class EmployeeServiceB {

    @Transactional
    public void operationB() {
    }
}
