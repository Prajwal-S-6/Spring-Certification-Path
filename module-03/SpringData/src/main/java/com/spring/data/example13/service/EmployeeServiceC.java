package com.spring.data.example13.service;

import org.springframework.transaction.annotation.Transactional;

@Transactional
public interface EmployeeServiceC {

    void operationC();
}
