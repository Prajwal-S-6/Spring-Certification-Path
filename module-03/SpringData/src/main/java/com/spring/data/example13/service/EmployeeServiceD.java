package com.spring.data.example13.service;

import org.springframework.transaction.annotation.Transactional;

public interface EmployeeServiceD {

    @Transactional
    void operationD();
}
