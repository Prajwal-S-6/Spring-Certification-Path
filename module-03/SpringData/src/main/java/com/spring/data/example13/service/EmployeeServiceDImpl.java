package com.spring.data.example13.service;

import org.springframework.stereotype.Service;

@Service
public class EmployeeServiceDImpl implements EmployeeServiceD {
    @Override
    public void operationD() {

    }
}
