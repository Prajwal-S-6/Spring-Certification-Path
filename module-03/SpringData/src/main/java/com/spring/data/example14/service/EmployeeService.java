package com.spring.data.example14.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import javax.sql.DataSource;
import java.sql.Connection;
import java.sql.SQLException;

@Service
public class EmployeeService {

    @Autowired
    @Qualifier("custom_data_source")
    private DataSource dataSource;

    @Transactional
    public void declarativeTransaction() {
        // business logic
    }

    public void manualTransaction() throws SQLException {
        Connection connection = dataSource.getConnection();

        connection.setAutoCommit(false);
        connection.setTransactionIsolation(Connection.TRANSACTION_READ_COMMITTED);

        try {
            // business logic

            connection.commit();
        } catch (SQLException e) {
            connection.rollback();
        } finally {
            connection.close();
        }
    }
}
