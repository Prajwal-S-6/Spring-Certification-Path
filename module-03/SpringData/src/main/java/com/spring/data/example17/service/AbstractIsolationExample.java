package com.spring.data.example17.service;


import com.spring.data.example17.dao.EmployeeDao;
import com.spring.data.example17.ds.Employee;
import org.springframework.beans.factory.annotation.Autowired;

import java.sql.Date;
import java.util.concurrent.CountDownLatch;

abstract class AbstractIsolationExample {
    @Autowired
    private EmployeeDao employeeDao;

    CountDownLatch fistReadOnUnsavedAndNotCommittedLatch = new CountDownLatch(1);
    CountDownLatch secondReadOnSavedButNotCommittedLatch = new CountDownLatch(1);
    CountDownLatch writeDoneNotCommittedLatch = new CountDownLatch(1);
    CountDownLatch writeDoneAndCommittedLatch = new CountDownLatch(1);
    CountDownLatch readWriteLatch = new CountDownLatch(2);

    void readOperation() {
        employeeDao.findAllEmployeesByIdRange(1, 6).forEach(System.out::println);
    }

    void writeOperation() {
        employeeDao.saveEmployee(new Employee(6, "Mary", "Duncan", "Mary.Duncam@cor.com", "555-55-60", Date.valueOf("2019-07-30"), 130000));
        employeeDao.maskEmail(3);
    }

    public CountDownLatch getFistReadOnUnsavedAndNotCommittedLatch() {
        return fistReadOnUnsavedAndNotCommittedLatch;
    }

    public CountDownLatch getSecondReadOnSavedButNotCommittedLatch() {
        return secondReadOnSavedButNotCommittedLatch;
    }

    public CountDownLatch getWriteDoneNotCommittedLatch() {
        return writeDoneNotCommittedLatch;
    }

    public CountDownLatch getWriteDoneAndCommittedLatch() {
        return writeDoneAndCommittedLatch;
    }

    public CountDownLatch getReadWriteLatch() {
        return readWriteLatch;
    }
}
