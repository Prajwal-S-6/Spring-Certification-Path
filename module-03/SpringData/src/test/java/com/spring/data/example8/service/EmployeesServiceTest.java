package com.spring.data.example8.service;

import ch.qos.logback.classic.Logger;
import ch.qos.logback.classic.spi.ILoggingEvent;
import ch.qos.logback.core.read.ListAppender;
import com.spring.data.example3.jdbc.template.service.TestDataSourceConfig;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.context.annotation.Import;
import org.springframework.test.context.ActiveProfiles;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;


@SpringBootTest
@Import(TestDataSourceConfig.class)
@ActiveProfiles("test")
class EmployeesServiceTest {

    @Autowired
    private EmployeesService employeesService;

    Logger logger = (Logger) LoggerFactory.getLogger(EmployeesService.class);

    @Test
    void shouldLogAllEmployees() {
        ListAppender<ILoggingEvent> listAppender = new ListAppender<>();
        logger.addAppender(listAppender);
        listAppender.start();

        List<ILoggingEvent> logs = listAppender.list;
        employeesService.printEmployees();

        Approvals.verify(logs.stream().map(ILoggingEvent::getFormattedMessage).reduce((log1, log2) -> log1 + "\n" + log2).get());

    }

    @Test
    void shouldDeleteAllEmployees() {
        ListAppender<ILoggingEvent> listAppender = new ListAppender<>();
        logger.addAppender(listAppender);
        listAppender.start();

        List<ILoggingEvent> logs = listAppender.list;
        employeesService.deleteAllEmployees();
        employeesService.printEmployees();

        Approvals.verify(logs.stream().map(ILoggingEvent::getFormattedMessage).reduce((log1, log2) -> log1 + "\n" + log2).get());

    }
}