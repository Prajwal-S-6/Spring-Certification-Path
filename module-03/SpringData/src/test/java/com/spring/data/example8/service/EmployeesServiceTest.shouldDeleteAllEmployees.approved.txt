Deleting all employees...
Printing list of employees